### PR TITLE
simd binary quantization and benches

### PR DESCRIPTION
**Pull request by IvanPleshkov**

### CODE REVIEW

File: quantization/src/encoded_vectors_binary.rs [hunk 1]
```diff
@@ -72,6 +72,29 @@ impl<TStorage: EncodedStorage> EncodedVectorsBin<TStorage> {
...
+        #[cfg(all(target_arch = "aarch64", target_feature = "neon"))]
+        if std::arch::is_aarch64_feature_detected!("neon") {
+            unsafe {
+                return impl_xor_popcnt_neon(
+                    v1.as_ptr() as *const u64,
+                    v2.as_ptr() as *const u64,
+                    v1.len() as u32,
+                ) as usize;
+            }
+        }
```
Comment (timvisee) [inline quantization/src/encoded_vectors_binary.rs]: For SSE4.1 we provide `count = 2 * v1.len()` and for Neon we provide `count = v1.len()`.

Function signatures are the same, though I see the implementations are different. Could we clarify this a bit with a comment, or by renaming one of  the `counts`?

**Review by timvisee — CHANGES_REQUESTED** [review]
Nice work!

I'm not familiar with Neon instructions at all so I didn't do very deep inspection.

The only thing I'd like to see is a unit test to assert `xor_product` output is the same as the default implementation. That would ensure that the SSE/Neon implementations are correct (when tested on those platforms).

It would literally be something like:

```rust
#[test]
fn test_xor_product() {
    // Have some test data here
    for (v1, v2) in vectors {
        assert_eq!(
            EncodedVectorsBin::xor_product(&v1, &v2),
            v1.iter()
                .zip(v2)
                .map(|(b1, b2)| (b1 ^ b2).count_ones() as usize)
                .sum()
        );
    }
}
```

File: quantization/src/encoded_vectors_binary.rs (hunk 1 above)
Comment (IvanPleshkov) [reply quantization/src/encoded_vectors_binary.rs]: Thanks for highlighting it! In arm, we process 128-bit chunks but in x64 only 64-bit, that's why multiplication by 2 is correct. But this is an implementation detail. We shouldn't know it. Fixed by moving this multiplication to C code

**Comment by IvanPleshkov** [comment]
> Nice work!
> 
> I'm not familiar with Neon instructions at all so I didn't do very deep inspection.
> 
> The only thing I'd like to see is a unit test to assert `xor_product` output is the same as the default implementation. That would ensure that the SSE/Neon implementations are correct (when tested on those platforms).
> 
> It would literally be something like:
> 
> ```rust
> #[test]
> fn test_xor_product() {
>     // Have some test data here
>     for (v1, v2) in vectors {
>         assert_eq!(
>             EncodedVectorsBin::xor_product(&v1, &v2),
>             v1.iter()
>                 .zip(v2)
>                 .map(|(b1, b2)| (b1 ^ b2).count_ones() as usize)
>                 .sum()
>         );
>     }
> }
> ```

I guess such test is not necessary. We have unit tests for binary quantization and each test covers xor. I would discuss testing of non-accelerated xor calculation but I decided to skip it. Because the implementation and test (like in your example) are absolutely the same parts of code

File: quantization/src/encoded_vectors_binary.rs (hunk 1 above)
Comment (timvisee) [reply quantization/src/encoded_vectors_binary.rs]: Fantastic!

**Review by timvisee — APPROVED** [review]

**Pull request merged**